File: shared/domain/src/main/java/com/ivy/domain/usecase/budget/DynamicBudgetEngine.kt
package com.ivy.domain.usecase.budget

import com.ivy.data.model.AllocationMode
import java.time.LocalDate

data class ExpenseAllocationEntry(
    val date: LocalDate,
    val amount: Long
)

interface ExpenseAllocationScheduler {
    fun schedule(
        amount: Long,
        transactionDate: LocalDate,
        mode: AllocationMode,
        periodEnd: LocalDate
    ): List<ExpenseAllocationEntry>
}

interface BudgetPeriod {
    val start: LocalDate
    val end: LocalDate
    fun daysRemaining(today: LocalDate): Int
}

data class BudgetExpenseInput(
    val transactionDate: LocalDate,
    val amount: Long,
    val mode: AllocationMode = AllocationMode.TODAY
)

data class BudgetEngineInput(
    val period: BudgetPeriod,
    val budgetLimit: Long,
    val includedIncome: Long = 0L,
    val reservationsTotal: Long = 0L,
    val expenses: List<BudgetExpenseInput> = emptyList(),
    val today: LocalDate
)

data class BudgetSnapshot(
    val capacity: Long,
    val periodSpentRaw: Long,
    val periodRemaining: Long,
    val daysRemaining: Int,
    val openingAllowance: Long,
    val todayCharges: Long,
    val remainingAllowance: Long,
    val tomorrowProjection: Long?
)

class DynamicBudgetEngine(
    private val scheduler: ExpenseAllocationScheduler
) {
    fun calculate(input: BudgetEngineInput): BudgetSnapshot {
        val period = input.period
        val today = input.today

        val daysRemaining = period.daysRemaining(today)
        val capacity = input.budgetLimit + input.includedIncome

        // Filter eligible expenses strictly within [period.start, period.end]
        val eligibleExpenses = input.expenses.filter { exp ->
            !exp.transactionDate.isBefore(period.start) && !exp.transactionDate.isAfter(period.end)
        }

        val periodSpentRaw = eligibleExpenses.sumOf { it.amount }
        val periodRemaining = capacity - periodSpentRaw - input.reservationsTotal

        // Build allocation map
        val allocationsMap = mutableMapOf<LocalDate, Long>()
        for (exp in eligibleExpenses) {
            val entries = scheduler.schedule(
                amount = exp.amount,
                transactionDate = exp.transactionDate,
                mode = exp.mode,
                periodEnd = period.end
            )
            for (entry in entries) {
                allocationsMap[entry.date] = (allocationsMap[entry.date] ?: 0L) + entry.amount
            }
        }

        // Charges strictly before today
        var chargesBeforeToday = 0L
        for ((date, amt) in allocationsMap) {
            if (date.isBefore(today)) {
                chargesBeforeToday += amt
            }
        }

        // Pool at start of today (00:00)
        val poolAtStartOfToday = capacity - input.reservationsTotal - chargesBeforeToday
        val openingAllowance = poolAtStartOfToday / maxOf(1, daysRemaining)

        val todayCharges = allocationsMap[today] ?: 0L
        val remainingAllowance = openingAllowance - todayCharges

        // Tomorrow projection: only when daysRemaining > 1 and today is before period.end
        val tomorrowProjection: Long? = if (daysRemaining > 1 && !today.isEqual(period.end)) {
            val daysFromTomorrow = daysRemaining - 1
            val effectiveTodayCharges = if (todayCharges < openingAllowance) openingAllowance else todayCharges
            val chargesBeforeTomorrow = chargesBeforeToday + effectiveTodayCharges
            val poolAtStartOfTomorrow = capacity - input.reservationsTotal - chargesBeforeTomorrow
            val tomorrowOpening = poolAtStartOfTomorrow / daysFromTomorrow

            val tomorrow = today.plusDays(1)
            val tomorrowCharges = allocationsMap[tomorrow] ?: 0L
            tomorrowOpening - tomorrowCharges
        } else {
            null
        }

        return BudgetSnapshot(
            capacity = capacity,
            periodSpentRaw = periodSpentRaw,
            periodRemaining = periodRemaining,
            daysRemaining = daysRemaining,
            openingAllowance = openingAllowance,
            todayCharges = todayCharges,
            remainingAllowance = remainingAllowance,
            tomorrowProjection = tomorrowProjection
        )
    }
}

File: src/domain/dynamicBudgetEngine.ts
import {
  BudgetSnapshot,
  DynamicBudgetConfig,
  BudgetPeriod,
  Transaction,
  Reservation,
} from '../types';
import { BudgetPeriodResolver } from './budgetPeriodResolver';
import { ExpenseAllocationScheduler } from './expenseAllocationScheduler';

export interface DynamicBudgetEngineInput {
  config: DynamicBudgetConfig;
  period: BudgetPeriod;
  transactions: Transaction[];
  reservations: Reservation[];
  todayStr: string; // YYYY-MM-DD
}

export class DynamicBudgetEngine {
  static calculate(input: DynamicBudgetEngineInput): BudgetSnapshot {
    const { config, period, transactions, reservations, todayStr } = input;

    if (config.budgetLimit <= 0) {
      return {
        periodStart: period.start,
        periodEnd: period.end,
        capacity: 0,
        periodSpentRaw: 0,
        periodReserved: 0,
        periodRemaining: 0,
        daysRemaining: BudgetPeriodResolver.daysRemaining(todayStr, period.end),
        openingAllowance: 0,
        todayCharges: 0,
        remainingAllowance: 0,
        tomorrowProjection: null,
        status: 'noBudget',
        overTodayAmount: 0,
      };
    }

    // Helper to check account and category eligibility
    const isAccountEligible = (accId?: string) => {
      if (!accId) return true;
      if (!config.includedAccountIds || config.includedAccountIds.length === 0) return true;
      return config.includedAccountIds.includes(accId);
    };

    const isCategoryEligible = (catId?: string) => {
      if (!catId) return true;
      if (!config.includedCategoryIds || config.includedCategoryIds.length === 0) return true;
      return config.includedCategoryIds.includes(catId);
    };

    // Calculate included income
    let includedIncome = 0;
    if (config.includeIncomeInBudget) {
      for (const tx of transactions) {
        if (
          tx.type === 'income' &&
          tx.includeInBudget &&
          tx.date.slice(0, 10) >= period.start &&
          tx.date.slice(0, 10) <= period.end &&
          isAccountEligible(tx.accountId) &&
          isCategoryEligible(tx.categoryId)
        ) {
          includedIncome += tx.amount;
        }
      }
    }

    const capacity = config.budgetLimit + includedIncome;

    // Filter active reservations in this period
    let periodReserved = 0;
    for (const res of reservations) {
      if (
        res.status === 'active' &&
        res.date >= period.start &&
        res.date <= period.end &&
        isAccountEligible(res.accountId) &&
        isCategoryEligible(res.categoryId)
      ) {
        periodReserved += res.amount;
      }
    }

    // Filter eligible expenses in this period
    const eligibleExpenses = transactions.filter((tx) => {
      if (tx.type !== 'expense') return false;
      const txDate = tx.date.slice(0, 10);
      if (txDate < period.start || txDate > period.end) return false;
      return isAccountEligible(tx.accountId) && isCategoryEligible(tx.categoryId);
    });

    const periodSpentRaw = eligibleExpenses.reduce((sum, tx) => sum + tx.amount, 0);
    const periodRemaining = capacity - periodSpentRaw - periodReserved;

    const daysRemaining = BudgetPeriodResolver.daysRemaining(todayStr, period.end);

    // Build allocation map
    const allocationsMap: Record<string, number> = {};
    for (const exp of eligibleExpenses) {
      const txDate = exp.date.slice(0, 10);
      const entries = ExpenseAllocationScheduler.schedule(
        exp.amount,
        txDate,
        exp.allocationMode || 'TODAY',
        period.end
      );
      for (const entry of entries) {
        allocationsMap[entry.date] = (allocationsMap[entry.date] || 0) + entry.amount;
      }
    }

    // Charges strictly before today
    let chargesBeforeToday = 0;
    for (const [date, amt] of Object.entries(allocationsMap)) {
      if (date < todayStr) {
        chargesBeforeToday += amt;
      }
    }

    // Pool at start of today (00:00)
    const poolAtStartOfToday = capacity - periodReserved - chargesBeforeToday;
    const isIntegerMath = Number.isInteger(capacity) && Number.isInteger(poolAtStartOfToday);
    const openingAllowance = isIntegerMath
      ? Math.trunc(poolAtStartOfToday / Math.max(1, daysRemaining))
      : poolAtStartOfToday / Math.max(1, daysRemaining);

    const todayCharges = allocationsMap[todayStr] || 0;
    const remainingAllowance = openingAllowance - todayCharges;

    // Tomorrow projection
    let tomorrowProjection: number | null = null;
    if (todayStr < period.end && daysRemaining > 1) {
      const daysFromTomorrow = daysRemaining - 1;
      const effectiveTodayCharges =
        todayCharges < openingAllowance ? openingAllowance : todayCharges;
      const chargesBeforeTomorrow = chargesBeforeToday + effectiveTodayCharges;
      const poolAtStartOfTomorrow = capacity - periodReserved - chargesBeforeTomorrow;
      const tomorrowOpening = isIntegerMath
        ? Math.trunc(poolAtStartOfTomorrow / daysFromTomorrow)
        : poolAtStartOfTomorrow / daysFromTomorrow;

      const tomorrowDate = new Date(todayStr + 'T00:00:00');
      tomorrowDate.setDate(tomorrowDate.getDate() + 1);
      const tomorrowStr = tomorrowDate.toISOString().slice(0, 10);
      const tomorrowCharges = allocationsMap[tomorrowStr] || 0;

      tomorrowProjection = tomorrowOpening - tomorrowCharges;
    }

    // Determine status
    let status: 'active' | 'overToday' | 'periodExhausted' | 'noBudget' = 'active';
    let overTodayAmount = 0;

    if (periodRemaining <= 0) {
      status = 'periodExhausted';
    } else if (remainingAllowance < 0) {
      status = 'overToday';
      overTodayAmount = Math.abs(remainingAllowance);
    }

    return {
      periodStart: period.start,
      periodEnd: period.end,
      capacity,
      periodSpentRaw,
      periodReserved,
      periodRemaining,
      daysRemaining,
      openingAllowance,
      todayCharges,
      remainingAllowance,
      tomorrowProjection,
      status,
      overTodayAmount,
    };
  }
}
